File: frontend/src/services/courseService.js
import { apiRequest } from "./apiClient";

function unwrap(response) {
  return response?.data;
}

export async function getCourses(params = {}) {
  const query = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== "") {
      query.set(key, value);
    }
  });
  const suffix = query.toString() ? `?${query}` : "";
  return unwrap(await apiRequest(`/courses${suffix}`, { skipAuth: true }));
}

export async function getCourseBySlug(slug) {
  return unwrap(await apiRequest(`/courses/${slug}`, { skipAuth: true }));
}

export async function getCourseChapters(courseId) {
  return unwrap(await apiRequest(`/courses/${courseId}/chapters`));
}

export async function getLesson(courseId, lessonId) {
  return unwrap(await apiRequest(`/courses/${courseId}/lessons/${lessonId}`));
}

export async function getCourseAccess(courseId) {
  return unwrap(await apiRequest(`/courses/${courseId}/access`));
}

export async function enrollFree(courseId) {
  return unwrap(await apiRequest(`/courses/${courseId}/enroll-free`, { method: "POST" }));
}

export async function getTeacherCourses(params = {}) {
  const query = new URLSearchParams(params).toString();
  return unwrap(await apiRequest(`/teacher/courses${query ? `?${query}` : ""}`));
}

export async function getTeacherCourseDetail(id) {
  return unwrap(await apiRequest(`/teacher/courses/${id}`));
}

export async function getTeacherCourseById(courseId) {
  return getTeacherCourseDetail(courseId);
}

export async function createCourse(payload) {
  return unwrap(await apiRequest("/teacher/courses", {
    method: "POST",
    body: JSON.stringify(payload),
  }));
}

export async function updateCourse(id, payload) {
  return unwrap(await apiRequest(`/teacher/courses/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  }));
}

export async function deleteCourse(id) {
  return unwrap(await apiRequest(`/teacher/courses/${id}`, { method: "DELETE" }));
}

export async function submitCourseReview(id) {
  return unwrap(await apiRequest(`/teacher/courses/${id}/submit-review`, { method: "POST" }));
}

export async function getTeacherCourseChapters(courseId) {
  return unwrap(await apiRequest(`/teacher/courses/${courseId}/chapters`));
}

export async function createChapter(courseId, payload) {
  return unwrap(await apiRequest(`/teacher/courses/${courseId}/chapters`, {
    method: "POST",
    body: JSON.stringify(payload),
  }));
}

export async function updateChapter(chapterId, payload) {
  return unwrap(await apiRequest(`/teacher/chapters/${chapterId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  }));
}

export async function deleteChapter(chapterId) {
  return unwrap(await apiRequest(`/teacher/chapters/${chapterId}`, { method: "DELETE" }));
}

export async function createLesson(chapterId, payload) {
  return unwrap(await apiRequest(`/teacher/chapters/${chapterId}/lessons`, {
    method: "POST",
    body: JSON.stringify(payload),
  }));
}

export async function updateLesson(lessonId, payload) {
  return unwrap(await apiRequest(`/teacher/lessons/${lessonId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  }));
}

export async function deleteLesson(lessonId) {
  return unwrap(await apiRequest(`/teacher/lessons/${lessonId}`, { method: "DELETE" }));
}

export async function getAdminCourses(params = {}) {
  const query = new URLSearchParams(params).toString();
  return unwrap(await apiRequest(`/admin/courses${query ? `?${query}` : ""}`));
}

export async function getAdminCourseDetail(id) {
  return unwrap(await apiRequest(`/admin/courses/${id}`));
}

export async function getAdminCourseChapters(courseId) {
  return unwrap(await apiRequest(`/admin/courses/${courseId}/chapters`));
}

export async function getAdminCourseReviewHistory(courseId) {
  return unwrap(await apiRequest(`/admin/courses/${courseId}/review-history`));
}

export async function approveCourse(id) {
  return unwrap(await apiRequest(`/admin/courses/${id}/approve`, { method: "POST", skipRefresh: true }));
}

export async function rejectCourse(id, reason = "") {
  return unwrap(await apiRequest(`/admin/courses/${id}/reject`, {
    method: "POST",
    skipRefresh: true,
    body: JSON.stringify({ reason }),
  }));
}

export async function publishCourse(id) {
  return unwrap(await apiRequest(`/admin/courses/${id}/publish`, { method: "POST", skipRefresh: true }));
}

export async function hideCourse(id) {
  return unwrap(await apiRequest(`/admin/courses/${id}/hide`, { method: "POST", skipRefresh: true }));
}

export async function archiveCourse(id) {
  return unwrap(await apiRequest(`/admin/courses/${id}/archive`, { method: "POST", skipRefresh: true }));
}

export async function updateCourseSale(id, payload) {
  return unwrap(await apiRequest(`/admin/courses/${id}/sale`, {
    method: "PUT",
    skipRefresh: true,
    body: JSON.stringify(payload),
  }));
}

export async function clearCourseSale(id) {
  return unwrap(await apiRequest(`/admin/courses/${id}/sale`, {
    method: "DELETE",
    skipRefresh: true,
  }));
}

File: backend/src/main/java/com/example/englishlearning/service/CourseService.java
package com.example.englishlearning.service;

import com.example.englishlearning.dto.common.PageResponse;
import com.example.englishlearning.dto.course.ChapterRequest;
import com.example.englishlearning.dto.course.ChapterResponse;
import com.example.englishlearning.dto.course.AdminCourseSaleRequest;
import com.example.englishlearning.dto.course.CourseAccessResponse;
import com.example.englishlearning.dto.course.CourseDetailResponse;
import com.example.englishlearning.dto.course.CourseRequest;
import com.example.englishlearning.dto.course.CourseReviewHistoryResponse;
import com.example.englishlearning.dto.course.CourseSummaryResponse;
import com.example.englishlearning.dto.course.LessonRequest;
import com.example.englishlearning.dto.course.LessonResponse;
import com.example.englishlearning.dto.course.RejectCourseRequest;
import com.example.englishlearning.entity.Course;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface CourseService {

    PageResponse<CourseSummaryResponse> getPublishedCourses(String search, Course.CourseLevel level, Course.CourseType courseType, Pageable pageable);

    CourseDetailResponse getPublishedCourseBySlug(String slug);

    List<ChapterResponse> getCourseChapters(Long courseId, String currentEmail);

    LessonResponse getLesson(Long courseId, Long lessonId, String currentEmail);

    CourseAccessResponse getAccess(Long courseId, String currentEmail);

    CourseAccessResponse enrollFree(Long courseId, String currentEmail);

    PageResponse<CourseSummaryResponse> getTeacherCourses(String teacherEmail, Pageable pageable);

    CourseDetailResponse getTeacherCourseDetail(Long courseId, String teacherEmail);

    CourseDetailResponse createCourse(CourseRequest request, String teacherEmail);

    CourseDetailResponse updateCourse(Long courseId, CourseRequest request, String teacherEmail);

    void deleteTeacherCourse(Long courseId, String teacherEmail);

    CourseDetailResponse submitReview(Long courseId, String teacherEmail);

    ChapterResponse createChapter(Long courseId, ChapterRequest request, String teacherEmail);

    ChapterResponse updateChapter(Long chapterId, ChapterRequest request, String teacherEmail);

    LessonResponse createLesson(Long chapterId, LessonRequest request, String teacherEmail);

    LessonResponse updateLesson(Long lessonId, LessonRequest request, String teacherEmail);

    void deleteChapter(Long chapterId, String teacherEmail);

    void deleteLesson(Long lessonId, String teacherEmail);

    List<ChapterResponse> getTeacherCourseChapters(Long courseId, String teacherEmail);

    PageResponse<CourseSummaryResponse> getAdminCourses(Course.CourseStatus status, Pageable pageable);

    CourseDetailResponse getAdminCourseDetail(Long courseId);

    List<ChapterResponse> getAdminCourseChapters(Long courseId);

    List<CourseReviewHistoryResponse> getAdminCourseReviewHistory(Long courseId);

    CourseDetailResponse approveCourse(Long courseId);

    CourseDetailResponse rejectCourse(Long courseId, RejectCourseRequest request);

    CourseDetailResponse publishCourse(Long courseId);

    CourseDetailResponse hideCourse(Long courseId);

    CourseDetailResponse archiveCourse(Long courseId);

    CourseDetailResponse upsertCourseSale(Long courseId, AdminCourseSaleRequest request);

    CourseDetailResponse clearCourseSale(Long courseId);
}
